pkg/plugin/velero/item_snapshotter/v1/item_snapshotter.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"time"
+
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+)
+
+type AlsoHandlesInput struct {
+	ItemToSnapshot runtime.Unstructured
+	Backup         *api.Backup
+}
+
+type SnapshotItemInput struct {
+	Item   runtime.Unstructured
+	Params map[string]string
+	Backup *api.Backup
+}
+
+type SnapshotItemOutput struct {
+	UpdatedItem      runtime.Unstructured
+	SnapshotID       string
+	SnapshotMetadata map[string]string // SnapshotMetadata is information in addition to the SnapshotID that the
+	// plugin wants to store in the backup.  SnapshotMetadata will be passed
+	// back in for DeleteSnapshot and CreateItemFromSnapshot
+	AdditionalItems []velero.ResourceIdentifier
+	HandledItems    []velero.ResourceIdentifier // Items that were handled by this snapshot that can be ignored
+}
+
+// ProgressInput contains the input parameters for the ItemSnapshotter's Progress function.
+type ProgressInput struct {
+	// ItemID is the id of item that was stored in the backup
+	ItemID velero.ResourceIdentifier
+	// SnapshotID is the snapshot ID returned by ItemSnapshotter
+	SnapshotID string
+	// Backup is the representation of the backup resource processed by Velero.
+	Backup *api.Backup
+}
+
+type ProgressOutput struct {
+	Completed                       bool      // True when the operation has completed, either successfully or with a failure
+	Err                             error     // Set when the operation has failed
```

Comment:
So this is the error if the snapshot processing failed.  If there's an error retrieving the progress that would be an error returned from the call.  I changed this to be a string since on further reflection returning even a JSON'd error would make the API too Go specific.  I also replaced Completed with a phase that reflects the state.